### ISSUE TITLE
Is it possible to support irregular grid in the ABF bias?

### ISSUE THREAD

**Issue by HanatoK** [Colvars/colvars#445]
I am thinking of running a 3D eABF simulation. However, only some regions in the 3D space are well-defined, and if eABF samples outside these regions the simulation is likely to crash. Is that possible to provide a way to tell the ABF bias only exerts forces on specific bins?
It occurs to me that it should be possible to make ABF accepts a list of grid files, and the grid files indicate whether ABF should collect and exert forces in specific bins. For example, if the user provides the following grid file:
```
# 3
# -1.000000000 0.010000000 200 0
# -1.000000000 0.010000000 200 0
# -1.000000000 0.010000000 200 0
      -0.99500       -0.99500       -0.99500      0
      -0.99500       -0.99500       -0.98500      0
      -0.99500       -0.99500       -0.97500      0
      -0.99500       -0.99500       -0.96500      0
      -0.99500       -0.99500       -0.95500      1
      -0.99500       -0.99500       -0.94500      0

```
then if the CV is in the bin of which the center is (-0.99500, -0.99500, -0.95500), the gradients are collected and the bias forces are applied since the bin center is marked as 1.

**Comment by giacomofiorin** [comment]
If the purpose is apply some confining restraints that cannot be defined simply with `harmonicWalls` my approach would be to just generate a 3D map that has a repulsive potential in regions that shouldn't be sampled.  There could be a new class added for that, but it wouldn't look very different from metadynamics or ABF with their respective options for updating the bias disabled.

Just to be clear, is this a a gimbal lock issue with the Euler angles, or do the "forbidden" regions of space depend on the specifics of the system?

**Comment by HanatoK** [comment]
@giacomofiorin I think the wall potentials can be solved by the scripted bias feature, but I want disable ABF forces on specific bins too.

My use case is for multidimensional CVs from ANN, which have irregular boundaries, and if the sampling goes far from the boundaries, the simulation may crash. An example of the boundaries is in the Figure 4 of Chen and Ferguson's paper (https://onlinelibrary.wiley.com/doi/10.1002/jcc.25520):
<img width="596" alt="Screenshot_20220121_092459" src="https://user-images.githubusercontent.com/24743682/150566315-6fbcbc63-7978-48eb-bcaa-9b4e9df472f0.png">
That being said, I want to disable/mask the biasing forces in bins marked as 0 in C.

**Comment by giacomofiorin** [comment]
That makes sense, thank you for clarifying.

In lieu of some Tcl restraints, how about initializing ABF not with an empty grid (default) but instead with a grid of gradients and samples prepared by you beforehand? This way you can "pretend" that ABF has sampled a landscape that gives you repulsive forces at the boundary bins, and you can set a really high number of samples at those bins (so that the initial estimate doesn't get averaged down by ABF). `colvar_grid_count` is based on `size_t`, so you can set fairly high numbers of samples.

Other approaches that involve new code would almost certainly need significant changes to the grid classes, which are more appropriate for a general overhaul than this specific use case.

**Comment by HanatoK** [comment]
@giacomofiorin Thanks, I can give the first idea a try.

**Comment by HanatoK** [comment]
@giacomofiorin I also want the metadynamics to apply the biasing forces in specific grids only. Is that possible? I don't think significant changes to the `colvargrid`. Maybe it should be OK to just add a parameter to `colvarbias` to accept an extra `colvargrid`, which filters the biasing force based on the grid values.

**Comment by HanatoK** [comment]
@giacomofiorin I made a draft PR #447 . Is that an feasible approach to scale the biasing force based on an external grid?

**Comment by giacomofiorin** [comment]
Closed by #447
